packages/x-date-pickers/src/CalendarPicker/CalendarPicker.tsx

```diff
@@ -113,6 +113,23 @@ export interface CalendarPickerProps<TDate>
    * @returns {boolean} If `true` the day will be disabled.
    */
   shouldDisableDate?: (day: TDate) => boolean;
+
+  /**
+   * Disable specific months dynamically.
+   * Works like `shouldDisableDate` but for month selection view @DateIOType.
+   * @template TDate
+   * @param {TDate} month The month to check.
+   * @returns {boolean} If `true` the month will be disabled.
+   */
+  shouldDisableMonth?: (month: TDate) => boolean;
+  /**
+   * Disable specific years dynamically.
+   * Works like `shouldDisableDate` but for year selection view @DateIOType.
+   * @template TDate
+   * @param {TDate} year The year to check.
+   * @returns {boolean} If `true` the year will be disabled.
+   */
+  shouldDisableYear?: (year: TDate) => boolean;
```

Comment:
We don't need to add `shouldDisableYear` here, it is already defined in `ExportedYearPickerProps`

And we could do the same for months with `ExportedMonthPickerProps`,`onMonthChange` and `shouldDisableMonth` and have `MonthPickerProps` extends it.